fpnge.cc

```diff
@@ -644,15 +641,19 @@ FORCE_INLINE void WriteBits(MIVEC nbits, MIVEC bits_lo, MIVEC bits_hi,
   // Merge bits_lo and bits_hi in 16-bit "bits".
   auto nbits0 = MM(unpacklo_epi8)(nbits, MMSI(setzero)());
   auto nbits1 = MM(unpackhi_epi8)(nbits, MMSI(setzero)());
-  auto bits_lo0 = MM(unpacklo_epi8)(bits_lo, MMSI(setzero)());
-  auto bits_lo1 = MM(unpackhi_epi8)(bits_lo, MMSI(setzero)());
-  auto bits_hi0 = MM(unpacklo_epi8)(bits_hi, MMSI(setzero)());
-  auto bits_hi1 = MM(unpackhi_epi8)(bits_hi, MMSI(setzero)());
+  MIVEC bits0, bits1;
+  if (mid_lo_nbits == 8) {
```

Comment:
IIRC that's the common case, right? In which case we probably want a LIKELY macro of some sort...

Reply:
I actually don't think it's even possible for it to be false, given how the Huffman table is constructed.

Even if you give it a 1px high PNG (which skips `CollectSymbolCounts` entirely) the supplied `kBaselineData` with the middle section having the lowest probability, always results in it getting 8 bits for the low half.

Even if that wasn't true, the value doesn't change throughout encoding, so it's easily predicted.

Reply:
Well if you can come up with any example where it isn't 8, I'm interested to hear it.

Of course, it being a variable allows for the *implementation* to be changed, so it makes sense to have it there, even if it's essentially a constant in the current design.

Reply:
Unfortunately that won't work, because the mid symbol allocation ignores symbol counts and [always sets it to minimal probability](https://github.com/veluca93/fpnge/blob/908faf438816ab4592d5ca1794e11b70c1c5ae37/fpnge.cc#L179). The low/high symbols (0-15, 240-255) probably will end up with ~5 bits each, but the middle will be stuck at 8.

Reply:
uhhhh I'm not sure it should be doing that...

I probably should clean up the code and add some tests at some point...